### ISSUE TITLE
What is the logic of Outlines? 

### ISSUE THREAD

**Issue by MMehdiMousavi** [Shreeyak/cleargrasp#17]
Hi, 
Im trying to make my own dataset for training, I don't understand why you are creating the outlines this way. 

as I understand it now, the outlines are rendered in grey (1,1,1) and where the bottles come into contact with the tray (or ground) and that is visible in the RGB image, it's a slightly lighter shade (2,2,2) in those areas.  but I do not understand why you are doing this. 

can you help me here? 

![comp](https://user-images.githubusercontent.com/30180265/105558057-ba697a00-5cdb-11eb-864a-d652b6ed7f49.png)

Best,

**Comment by Shreeyak** [comment]
The global optimization module which predicts the depth uses occlusion boundaries as an input. To help the edge prediction model learn better, we added an extra class of contact edges. 
The outlines are predicted through semantic segmentation, where each pixel is assigned a class. There are 2 types of edges: occlusion boundaries (which indicate a large differential on the depth) and the contact edges (edges in contact with other surfaces, and hence do not have the large differential).
In the paper, we visualize the 3 classes (background, occlusion, contact) through the colors red, green, blue.

**Comment by MMehdiMousavi** [comment]
Okay, but is this calculated when synthesizing the image? if so, why the neck of the bottle is classified as contact edge even though it's not really touching the tray?

**Comment by Shreeyak** [comment]
These were calculated offline after rendering the images. I went for a strategy of choosing a small threshold for depth gradient, but that gave me some incorrect values because some of the bottle are not actually touching the bottom surface. This is due to the physics simulation we ran when rendering - the bottles were simulated using a coarse convex hull of the object.

To overcome that, I made all edges below a certain height off the ground as contact edges. Even though that's technically incorrect, it was the best compromise at the time.

For your own data, if the objects are correctly sitting on the ground, applying a gradient to the depth image would be the way to go.

**Comment by MMehdiMousavi** [comment]
Thanks for the reply. 
One more question, can you explain what is saved in the Json files? and how I can recreate them?

**Comment by Shreeyak** [comment]
I'm sorry, which json? Could you share the contents of the json here.
If it's part of the dataset, those were generated by our rendering script. Unfortunately, the rendering script is not released.

**Comment by MMehdiMousavi** [comment]
Hi, thanks for the reply. 
there is a JSON file for each of the images that describes the rotation and location of each object. I cant see their use in the training code, I'm saying that because I managed to train a surface normal model without them in my data. 
in the paper, you mention its part of the dataset. If you tell me _where_ its used would be also a great help. since in my image synthesis pipeline it's very cumbersome to acquire them in this format.

here is one of them: https://pastebin.com/QjNSZV2e

**Comment by Shreeyak** [comment]
The Jsons are used to generate the segmentation masks (although not strictly necessary, since we only have 2 classes: background/transparent). They also contain the camera intrinsics.
Apart from that, they're not really used. It's there as additional metadata in case someone needs it.

**Comment by Shreeyak** [comment]
Update: Found the script used for creating the outlines. Have added it to the repo here: https://github.com/Shreeyak/cleargrasp/blob/master/z-ignore-scripts-helper/data_processing_script.py
